### PR TITLE
MAILBOX-283 Introduce memory implementation of MessageIdManager

### PR DESCRIPTION
**Pull request by aduprat**
Only MAILBOX-283 related commits.

This memory implementation of MessageIdManager is a wrapper of MessageManager, so it's sending events as StoreMessageManager not as StoreMessageIdManager.
This is why there are no SideEffects tests implementation.

### CODE REVIEW

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/InMemoryMessageId.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    public String toString() {
+        return String.valueOf(value);
+    }
+
+    public long getRawId() {
+        return value;
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (chibenwa) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/InMemoryMessageId.java]: Guava ?

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/InMemoryMessageId.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + (int) (value ^ (value >>> 32));
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (chibenwa) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/InMemoryMessageId.java]: Idem

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/InMemoryMessageIdManager.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    @Override
+    public void setInMailboxes(MessageId messageId, List<MailboxId> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
+        List<MessageResult> messages = getMessages(ImmutableList.of(messageId), FetchGroupImpl.MINIMAL, mailboxSession);
+        if (!messages.isEmpty()) {
+            ImmutableSet<MailboxId> currentMailboxes = currentMailboxes(messages).toSet();
+            
+            HashSet<MailboxId> targetMailboxes = Sets.newHashSet(mailboxIds);
+            List<MailboxId> mailboxesToRemove = ImmutableList.copyOf(Sets.difference(currentMailboxes, targetMailboxes));
```
Comment (chibenwa) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/InMemoryMessageIdManager.java]: Ok, I did not understand it this way... My bad. Some changes are required on https://github.com/linagora/james-project/pull/507

(Unsure that we remove from mailbox)

**Comment by chibenwa** [comment]
BTW, fix equals, and consider me as a :+1:

**Comment by aduprat** [comment]
test this please

File: protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java [hunk 4]
```diff
@@ -152,6 +153,11 @@ public char getHierarchyDelimiter() {
                     public MailboxPath getPath() {
                         return rootPath;
                     }
+
+                    @Override
+                    public MailboxId getId() {
+                        return null; //Will not be call in ListProcess scope
```
Comment (rouazana) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java]: *ListProcessor

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/json/JacksonEventSerializer.java [hunk 5]
```diff
@@ -57,38 +59,40 @@ public JacksonEventSerializer(EventConverter eventConverter, ObjectMapper object
...
 
-    public static ObjectMapper configureObjectMapper(ObjectMapper objectMapper) {
+    public static ObjectMapper configureObjectMapper(ObjectMapper objectMapper, MessageId.Factory messageIdFactory) {
         SimpleModule module = new SimpleModule();
         module.addDeserializer(MessageUid.class, new MessageUidDeserializer());
         module.addKeyDeserializer(MessageUid.class, new MessageUidKeyDeserializer());
         module.addSerializer(MessageUid.class, new MessageUidSerializer());
         module.addKeySerializer(MessageUid.class, new MessageUidKeySerializer());
+        module.addSerializer(MessageId.class, new MessageIdSerializer());
+        module.addDeserializer(MessageId.class, new MessageIdDeserializer(messageIdFactory));
```
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/json/JacksonEventSerializer.java]: maybe you could add some tests to EventSerializerTest?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageResult.java [hunk 6]
```diff
@@ -264,4 +264,6 @@
          */
         int[] getPositions();
     }
+
+    MessageId getMessageId();
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageResult.java]: you could change the commit title

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/InMemoryMessageIdManagerTestSystem.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            return messageManager.appendMessage(new ByteArrayInputStream("Subject: test\r\n\r\ntestmail".getBytes()), new Date(), mailboxSession, false, flags)
+                    .getMessageId();
+        } catch (MailboxException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    @Override
+    public MessageId createNotUsedMessageId() {
+        return InMemoryMessageId.of(RANDOM.nextLong());
```
Comment (rouazana) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/InMemoryMessageIdManagerTestSystem.java]: please, no random in tests

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/MessageIdManagerTestSystemProvider.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            mailboxManager.createMailbox(mailboxPath, mailboxSession);
+            MessageManager messageManager = mailboxManager.getMailbox(mailboxPath, mailboxSession);
+            return new SimpleMailbox(mailboxPath, randomUidValidity(), messageManager.getId());
+        } catch (MailboxException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    private static int randomUidValidity() {
+        return Math.abs(RANDOM.nextInt());
```
Comment (rouazana) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/MessageIdManagerTestSystemProvider.java]: idem

File: protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java (hunk 4 above)
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java]: Maybe you can even trhow an exception here no ? So we are really sure it's actually never called, and we can even avoid this comment

**Comment by aduprat** [comment]
#511

**Pull request closed without merging**